### ISSUE TITLE
Some questions about sequence parallel.

### ISSUE THREAD

**Issue by yingtongxiong** [Dao-AILab/flash-attention#625]
Hello, I recently implemented the sequence parallel in megatron by using flash-attn. I found a all-gather communicator in the ColumnParallelLinear, which is correct. However, I am very confused about the backward. For example, the input of layernorm is split in sequence dimension, which is like data parallel. Therefore, there needs an all-reduce operation in the layernorm to make the weight the same. I do not know where the all-reduce is. Maybe the reduce-scatter operation in the backward of ColumnParallelLinear?

**Comment by tridao** [comment]
Yes you need the all-reduce to sync the layer norm gradient / weights. That's part of the training loop, not part of the model implementation (or rather I couldn't find a good way to make that part of the model impl). Megatron has some [function](https://github.com/NVIDIA/Megatron-LM/blob/feac76a79148622d8f2a45d46c08a972a24784a3/megatron/core/pipeline_parallel/distrib_grad.py#L67) for this that gets called after the backward pass I think.

**Comment by yingtongxiong** [comment]
Okay, thank you very much.
